src/base/base.c

```diff
@@ -1067,6 +1123,79 @@ static zend_observer_fcall_handlers xdebug_observer_init(zend_execute_data *exec
 	return (zend_observer_fcall_handlers){xdebug_execute_begin, xdebug_execute_end};
 }
 #endif
+
+static void xdebug_enable_statement_handlers()
+{
+#if PHP_VERSION_ID >= 80100
+	zend_execute_ex = xdebug_execute_ex;
+#endif
+	
+	enable_statement_extension_handler();
+}
+
+static void xdebug_disable_statement_handlers()
+{
+#if PHP_VERSION_ID >= 80100
+	zend_execute_ex = xdebug_old_execute_ex;
+#endif
+	
+	disable_statement_extension_handler();
+}
+
+static void add_stack_frame_recursively(zend_execute_data *execute_data)
+{
+	zend_op_array *op_array;
+	int type;
+	function_stack_entry *fse;
+
+	if (execute_data->prev_execute_data) {
+		add_stack_frame_recursively(execute_data->prev_execute_data);
+	}
+	if (execute_data->func) {
+		op_array = &(execute_data->func->op_array);
+		type = ZEND_USER_CODE(execute_data->func->type) ? XDEBUG_USER_DEFINED : XDEBUG_BUILT_IN;
+		fse = xdebug_add_stack_frame(execute_data, op_array, type, 0);
```

Comment:
We use the xdebug_add_stack_frame() function to build the stack frame in the same way as when we normally run through the code